drivers/at_cmd/at_cmd.c

```diff
@@ -0,0 +1,293 @@
+		last_error = atoi(&buf[ARRAY_SIZE(AT_CMD_CMS_STR) - 1]);
+	} else {
+		last_error = 0;
+	}
+
+	if (tmpstr) {
+		buf[tmpstr - buf] = '\0';
+	}
+
+	return return_code;
+}
+
+static void callback_worker(struct k_work *item)
+{
+	struct callback_work_item *data =
+		CONTAINER_OF(item, struct callback_work_item, work);
+
+	if ((data != NULL) && (data->callback != NULL)) {
+		data->callback(data->data);
+	}
+
+	k_free(data);
+}
+
+
+static void socket_thread_fn(void *arg1, void *arg2, void *arg3)
+{
+	int bytes_read;
+	char buf[RESPONSE_MAX_LEN];
+	enum at_cmd_rc return_code;
+	bool write_with_response = false;
+
+	ARG_UNUSED(arg1);
+	ARG_UNUSED(arg2);
+	ARG_UNUSED(arg3);
+
+	LOG_DBG("AT socket thread started");
+
+	for (;;) {
+		bytes_read = recv(common_socket_fd, buf, sizeof(buf), 0);
+		write_with_response = false;
+
+		if (bytes_read < 0) {
+			LOG_ERR("AT socket recv failed with err %d",
+				bytes_read);
+
+			if (errno == -EAGAIN) {
+				if (open_socket() == 0) {
+					return_code = AT_CMD_ERROR;
+					goto next;
+				}
+			}
+
+			LOG_ERR("Unrecoverable reception error (err: %d),"
+				"thread killed", errno);
+			return;
+		} else if (bytes_read == sizeof(buf) ||
+			   buf[bytes_read - 1] != '\0') {
+
+			LOG_ERR("AT message to large for reception buffer or "
+				"missing termination character");
+			return_code = AT_CMD_ERROR;
+			goto next;
+		}
+
+		return_code = get_return_code(buf);
+
+		if (return_code != AT_CMD_NOTIFICATION) {
+			if ((response_buf_len > 0) &&
+			    (response_buf != NULL)) {
+				if (response_buf_len > strlen(buf)) {
+					strncpy(response_buf,
+						buf,
+						response_buf_len);
+				} else {
+					LOG_ERR("Response buffer not large "
+						"enough");
+					response_buf = NULL;
+				}
+
+				write_with_response = true;
+				response_buf_len    = 0;
+				response_buf        = NULL;
+			}
+		}
+
+		if (strlen(buf) == 0) {
+			goto next;
+		}
+
+		if (!write_with_response) {
+			struct callback_work_item *item =
+			k_malloc(sizeof(struct callback_work_item));
+
+			if (!item) {
+				LOG_DBG("Failed to allocate work item");
+				goto next;
+			}
+
+			k_work_init(&item->work, callback_worker);
+
+			memcpy(item->data, buf, strlen(buf) + 1);
+
+			if (return_code == AT_CMD_NOTIFICATION) {
+				item->callback = notification_handler;
+			} else {
+				item->callback = current_cmd_handler;
+			}
+
+			k_work_submit(&item->work);
+		}
+next:
+		if (return_code != AT_CMD_NOTIFICATION) {
+			k_msgq_put(&return_code_msq, &return_code, K_FOREVER);
+		}
+		continue;
+	}
+}
+
+static inline enum at_cmd_rc at_write(const char *const cmd)
+{
+	int bytes_sent;
+	enum at_cmd_rc return_code;
+
+	LOG_DBG("Sending command %s", cmd);
+	bytes_sent = send(common_socket_fd, cmd, strlen(cmd), 0);
+
+	if (bytes_sent < 0) {
+		LOG_ERR("Failed to send AT command");
+	} else {
+		k_msgq_get(&return_code_msq, &return_code, K_FOREVER);
+	}
+
+	LOG_DBG("Bytes sent: %d", bytes_sent);
+
+	return return_code;
```

Comment:
If it fails with errno == -EAGAIN it will return AT_CMD_ERROR, if it fails iwith errno != -EAGAIN the reception thread will die. But maybe I should send a message to the message queue so I don't stall the thread calling the write function, and then it should probably be a new one to indicate that something went horrible wrong.

Reply:
I was thinking about the situation when `bytes_sent < 0`, in that case reception thread has nothing to do with `return_code` value, it is returned as it was initialized (which will be an unspecified value in this case).

Reply:
aha, sorry, I read you comment incorrectly. You where talking about the at_write function and the return_code there, while I was looking at the return_code mechanism inside the thread. 

Hmm, I think I would like to initialize the variable to some value, the user must be able to detect that something went wrong with the sending even in the situation where the thread is not in involved, ie. when send fails.  So I was thinking that AT_CMD_ERROR is pretty generic so I though I can just initialize the variable to that. I don't think the user application needs to know  where it went wrong, just that it went wrong. In a debug situation the log output will anyway identify where it went wrong.